### ISSUE TITLE
Expose opera's TOSCA parser as a separate component

### ISSUE THREAD

**Issue by anzoman** [xlab-si/xopera-opera#170]
## Description
Just an idea, but maybe reality someday... Our xOpera orchestration tool has its own [TOSCA YAML parser](https://github.com/xlab-si/xopera-opera/tree/master/src/opera/parser) that gets the necessary orchestration data from TOSCA templates and uses it within the implementations (Ansible playbooks) to deploy the solutions. To extend the usage of our TOSCA parser we could move it to a separate repository to manage it as a separate entity of xOpera's project and then expose it as a component that would be primarily used within opera orchestrator, but can also be used by other tools that need TOSCA parsing functionality. This component could be available as a new Python package published on PyPI (`opera-tosca-parser` maybe) and it can be then installed and imported for usage within Python files.

## Steps
TBD if this will be chosen for implementation.

## Current behaviour
The `opera` orchestrator has embedded TOSCA parser that can already be used by installing and importing the opera python package.

## Expected results
Separate parser from `opera` and offer it as a separate reusable component.

**Comment by anzoman** [comment]
Let's elaborate this - as we will be adding support for [TOSCA 2.0](https://docs.oasis-open.org/tosca/TOSCA/v2.0/TOSCA-v2.0.html) this could be a good moment to separate TOSCA parser component from the xOpera orchestrator and make it available as a new `opera-tosca-parser` PyPI package (similar to our [`opera-api`](https://pypi.org/project/opera-api/) and also take look at [OpenStack TOSCA Parser](https://github.com/openstack/tosca-parser) - [`tosca-parser`](https://pypi.org/project/tosca-parser/)). But this would bring good and bad:

**Pros**:
This would reduce the size of the orchestrator's code and the parser would be added as a Python requirement. This way the orchestrator would focus only on managing data flow, the executors (e.g., Ansible, bash, Python) behind TOSCA. This way we could also extend our TOSCA parser component with multiple TOSCA versions (1.0, 1.1, 1.2, 1.3, 2.0) and then use these in the orchestrator. This is also good because the users, who seek just parsing would be able to use just our parser instead of downloading the whole orchestrator. By having the parser pluggable and separated from the orchestrator, this also brings a good setup for implementing and using other types of parsers that are not TOSCA - using specifications like OpenStack Heat, AWS CloudFormation, Terraform HCL and others. These parsers could be then developed separately from the core orchestrator which also splits the work better and also our issues and PRs would be clearer.

**Cons**:
We will have another project to manage, release and test. We would also need to document the TOSCA parser better (which is not a bad thing because the users would see how parsing is done and which TOSCA types are supported). As we will improve our parser, we will release new versions, which means that we will also have to constantly update the version within requirements.

@cankarm, @sstanovnik, @alexmaslenn, @dradx what do you think about this?

**Comment by alexmaslenn** [comment]
Having a second thought, I think it may be a good idea. The hardest part would be, if we really want to create a TOSCA independent parser in the future, is to create a universal infrastructure topology metamodel, as current xOpera implementation relies heavily on TOSCA 1.3 entity model.

My question is, are there really any use-cases when users seek just TOSCA parsing? I mean even Openstack parser creates a Heat Orchestration Template, which has application by its own. What would be an output of any TOSCA parser component? I see it more as a con, really, as we would have a Python package that would have zero application by its own.

**Comment by cankarm** [comment]
@alexmaslenn the output of TOSCA parser is a template; I guess that it should be a topology tree in some way, which could be used not only by our executors that initiate the deployment, but also by other uses who would like to:
* try only parsing of the TOSCA at some point and use our parser to do it (e.g. could be used in a process used, e.g., when TOSCA stuff is uploaded in Template Library Publishing System), 
* maybe leverage from the parsed model for other purposes, maybe to make a visual representation of the topology?
* other...
 
 So, I guess that there are some applications possible.
 
 If this does not take too much time, probably it is OK to proceed.

**Comment by anzoman** [comment]
@alexmaslenn I remember that similar functionality has been sought after here: https://github.com/radon-h2020/xopera-opera/issues/7 and probably there will be more. The output of our parser component with be similar to ours in `opera` now: an abstract syntax tree in the form of a Python `dict` or [ServiceTemplate](https://github.com/xlab-si/xopera-opera/blob/master/src/opera/parser/tosca/v_1_3/service_template.py#L21) object.

@cankarm thanks, I will proceed with this eventually and I have in mind that we keep the TOSCA 1.3 in our opera TOSCA parser  and then gradually add TOSCA 2.0, which will become the preferred (default) version in `opera`, but the users of `opera-tosca-parser` package will still be able to use 1.3 if needed (maybe we could even offer this downgrade from 2.0 to 1.3 in opera if ever needed or at the beginning we could support both versions).

**Comment by anzoman** [comment]
The new repository for parser component https://github.com/xlab-si/xopera-tosca-parser has been created.

**Comment by alexmaslenn** [comment]
@anzoman the code look good, I've run a number of tests, did not find discrepancies with `validate` command. As we are planning on using it as a separate component in opera, I guess `parse_service_template` and `parse_csar` should have a return value (a topology template probably). 

The main question is how should we connect parser to orchestrator? I mean after template is parsed by the parser it must be instantiated by orchestrator. Any thoughts on that part?

**Comment by anzoman** [comment]
@alexmaslenn cool, I have also connected `parse_service_template` and `parse_csar` functions into `parse` for easier calling. So, yes probably we should return `Topology` object which can then be used by here in `opera`. I have already pushed new commits for that and will release a new version.

So, yes the parser could be connected by using those three functions above and obtaining `Topology` object, which can then be instantiated. We can also use other functions from the parser such as `tosca.load`, `get_template` and others if we see that only three aforementioned functions are not enough.
